### ISSUE TITLE
issue when a person leaves the room?

### ISSUE THREAD

**Issue by greko6** [Dirvann/mediasoup-sfu-webrtc-video-rooms#2]
Hi Dirvann!

I was just testing around, nice implementation!

It looks like all consumers are freezing at the same time when a person joins or leaves the room. It's especially clear where there are 3+ people connected. You'll not see/notice it with two people. Do you know why it might be the case?

**Comment by greko6** [comment]
BTW to reproduce: connect with 3 clients, and just refresh the page in one of them while observing 2 remaining streams in 2 other browsers. They will hang and maybe reconnect a bit later. 

I did find the root cause. It happens when Peer transport is being closed(), here:

https://github.com/Dirvann/mediasoup-sfu-webrtc-video-rooms/blob/35ff76514b025663bef114c7fc3ad24f782d53fe/src/Peer.js#L101

What would be the reason that suddenly other peers that are in the room are disconnected as well?

**Comment by Dirvann** [comment]
Hi,  I've tried your setup with 3 browsers having video and audio share on.  I refreshed one page and everything worked as expected. The stream has been removed from the two remaining browsers and they can still see eachother.
I have tested this on chrome and serving with ubuntu.

Are you using another browser maybe?

When the peer disconnects it can call socketio disconnect https://github.com/Dirvann/mediasoup-sfu-webrtc-video-rooms/blob/35ff76514b025663bef114c7fc3ad24f782d53fe/src/app.js#L206 
Which calls [Room.removePeer(peer)](https://github.com/Dirvann/mediasoup-sfu-webrtc-video-rooms/blob/35ff76514b025663bef114c7fc3ad24f782d53fe/src/Room.js#L121) to remove the peer that has disconnected.
This closes the transports as you stated, closes the producers and sends an event to the consumers connected to the producer (which turns off & removes the stream from the closed browser)

**Comment by greko6** [comment]
Hmm... that's interesting. I'm using Chrome as well, connected to my Ubuntu (not VM) that runs this instance. 

It basically looks like when the stream is interrupted (i.e. disconnect event is triggered), all of the remaining streams are disconnected as well. 

Did you try to open the same stream from the same browser? (in my case: one incognito  + regular chrome + some other browser). 

I wonder what's different, as when I run exactly the same experiment with the same browser and OS, when I go to https://v3demo.mediasoup.org/ I don't have this behavior at all. Maybe they do run a devel branch of mediasoup that doesn't have this issue?

**Comment by Dirvann** [comment]
I updated mediasoup and added extra logs. 
I do notice that in the recent version of firefox (79.0) it seems to have a webrtc issue. Other than that I reproduced the configuration with chrome (Version 84.0.4147.125), chrome incognito and opera in different orders and I haven't had any issues. 
What do the logs say?

**Comment by greko6** [comment]
Heh, I went back to see if there was any update and saw your comment :) Apologies for the delay.

Here is the output. The issue starts at ### comment I've put below.

`
(...)
---produce--- type: video name: bob768 id: d44c3f5e-2c04-4a67-84ed-f140c8f11a80
---consuming--- name: bob321 prod_id:d44c3f5e-2c04-4a67-84ed-f140c8f11a80 consumer_id:9c909468-e8fe-4d22-8441-03b830fa249d
---user joined--- "123": bob885
---get RouterRtpCapabilities--- name: bob885
---create webrtc transport--- name: bob885
---adding transport--- 9b27d535-ba29-4c7c-8fb3-24c9eab75fb7
---create webrtc transport--- name: bob885
---adding transport--- d38cd2df-99c5-470b-9877-285dcdbe751e
---get producers--- name:bob885
---consuming--- name: bob885 prod_id:879bdf7e-01ac-451f-acf6-0f441e6ffe93 consumer_id:875da1a6-afcd-4334-afe5-adc92350864b
---consuming--- name: bob885 prod_id:d44c3f5e-2c04-4a67-84ed-f140c8f11a80 consumer_id:747b19c7-c5a8-4aad-9b7c-48f29c8dad2a
---connect transport--- name: bob885
---connect transport--- name: bob885
---produce--- type: video name: bob885 id: 92f40a87-e7b8-458d-ad12-f2bb33194748
---consuming--- name: bob321 prod_id:92f40a87-e7b8-458d-ad12-f2bb33194748 consumer_id:fd3245a5-5cad-4392-82d1-62c07d82008b
---consuming--- name: bob768 prod_id:92f40a87-e7b8-458d-ad12-f2bb33194748 consumer_id:eb62be21-24f7-43bb-a14c-e51f6b41473b

### DISCONNECTION STARTS HERE: just refreshed one of the browsers, immediately all of the transports are closing (instead of just one?)
---disconnect--- name: bob321
---producer transport close--- name: bob321 consumer_id: 879bdf7e-01ac-451f-acf6-0f441e6ffe93
---consumer transport close--- name: bob321 consumer_id: 9c909468-e8fe-4d22-8441-03b830fa249d
---consumer transport close--- name: bob321 consumer_id: fd3245a5-5cad-4392-82d1-62c07d82008b
---consumer closed--- due to producerclose event  name:bob885 consumer_id: 875da1a6-afcd-4334-afe5-adc92350864b
---consumer closed--- due to producerclose event  name:bob768 consumer_id: ac4e0993-195d-4d0b-bc67-5ddcce0ab2f1
`

At that point, all of the remaining peers are frozen. Do you know what this might be?

Browser Chrome: 85.0.4183.102 (Official Build) (64-bit) (cohort: Stable) 

Configuration: This git repo running inside docker, with udp/tcp ports exposed on a range of 44000-441000

**Comment by Dirvann** [comment]
- I don't know why you opened ports 44000-44100. The ports that have to be opened are 3016 (port of page) and 10000-10100 (port of rtc connection). unless you changed those ports.


The behavior of the server is normal. You have 3 users (bob321, 885 and 768). 
As 321 closes. its producer and 2 consumers (from other users) close because the transport has closed (due to disconnect).
Now the consumers of 885 and 768 were listening to the producer of 321 that just closed. So those two consumers also close.
So no problem there.

As the consumers of 885 and 768 close, it sends a message to the browser to remove the consumer. And the browser should stop the track on the two remaining browsers and remove the video element. 

https://github.com/Dirvann/mediasoup-sfu-webrtc-video-rooms/blob/2cfb6bd635faf8da2cc3a431c4b20005437bc9eb/src/Room.js#L112

https://github.com/Dirvann/mediasoup-sfu-webrtc-video-rooms/blob/2cfb6bd635faf8da2cc3a431c4b20005437bc9eb/public/RoomClient.js#L215

Also tried it on docker and that worked. 
I don't know what you mean by "all the peers are frozen".

**Comment by greko6** [comment]
Thank you for your reply.

Yes, I had to change the ports, but I did update the config.

We have 3 members that are producing and consuming. All of the members are cross-connected. When a member, let's call him "2" disconnects, then a video between a member "1" and "3" are freezing permanently (or if lucky, until a reconnection attempt). It's like, transport.close() is disconnecting all of them, where it should just disconnect gracefully a member "2". Hope it makes sense?

If a member "2" disconnect, I think a connection between "1" and "3" should not be affected... but for some reason it is, and I have just no idea why. Is it because all 3 members are coming from the same IP, and transport.close() is killing all connections or the browser doesn't track connected instances properly?
